tensorflow_datasets/core/custom_dataset/image_folder.py

```diff
@@ -0,0 +1,181 @@
+# coding=utf-8
+# Copyright 2020 The TensorFlow Datasets Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Lint as: python3
+"""Image Classification Folder datasets."""
+
+import os
+from typing import Dict, List, Tuple, Set
+
+import tensorflow.compat.v2 as tf
+from tensorflow_datasets import core
+
+# Dict of `split_label_images['split']['Label']` = `List['images']`
+SplitLabelImageDict = Dict[str, Dict[str, List[str]]]
+
+_SUPPORTED_IMAGE_FORMAT = ('.jpg', '.jpeg', '.png')
+
+
+class ImageFolder(core.DatasetBuilder):
+  """Generic image classification dataset created from manual directory.
+
+  The data directory should have the following structure:
+
+  ```
+  path/to/manual_dir/<dataset_name>/
+    split_name/  # Ex: 'train'
+      label1/  # Ex: 'airplane' or '0015'
+        xxx.png
+        xxy.png
+        xxz.png
+      label2/
+        xxx.png
+        xxy.png
+        xxz.png
+    split_name/  # Ex: 'test'
+      ...
+  ```
+
+  To use it:
+
+  ```
+  builder = tfds.ImageFolder(root_dir='path/to/manual_dir/')
+  print(builder.info)  # Splits, num examples,... are automatically calculated
+  ds = builder.as_dataset(split='split_name', shuffle_files=True)
+  tfds.show_examples(ds, builder.info)
+  ```
+
+  """
+
+  VERSION = core.Version('2.0.0')
+
+  def __init__(self, root_dir, **kwargs):
+    del kwargs
```

Comment:
We get extra args like `data_dir`, `config` in tests.
I added `kwargs` here because I could not find a way to handle these extra args by `functools`.

Reply:
If this is only for tests, then the test init function should be patched, not the original file.
If functools do not works, then use a function.